ssg/boolean_expression.py

```diff
@@ -69,14 +69,17 @@ class Symbol(boolean.Symbol):
 
     def __init__(self, obj):
         super(Symbol, self).__init__(obj)
-        self.spec = pkg_resources.Requirement.parse(obj)
-        self.obj = self.spec
+        self.requirement = pkg_resources.Requirement.parse(obj)
+        self.obj = self.requirement
 
     def __call__(self, **kwargs):
-        val = kwargs.get(self.name, False)
-        if len(self.spec.specs):
+        full_name = self.name
```

Comment:
This code is mixes a lot of abstraction levels together.
It is not clear why the object representing a parsed platform unit specification should be callable in the first place, and the underlying code is convoluted.

Reply:
This code know nothing about platforms, CPEs etc. It is an expansion of boolean.py's classes Function and Symbol which understand and can operate on boolean expressions that contain pkg_resources.Requirement as individual elements. The __init__, __call__ and __lt__ overload is essential for those symbols to be correctly compared, sorted and optimized, this is how boolean.py operates.

Reply:
Would it then be possible to represent the parser and parsed results as different classes? The current approach violates the interface segregation principle, and implementation bits of parsers are leaked to the code that is only interested in trees or platform units.